### PR TITLE
Avoid waypoint following during backtransition which can lead to strong banking

### PR DESCRIPTION
**Pull request by RomanBapst**
### Solved Problem
When the vehicle overshoots the transition, the controller tries to fly back to the current waypoint and commands a large banking angle. This is not desired during a backtransition as it can cause the transition to be prolonged and also it will go off path which is most likely not desired.

### Solution
During the backtransition let the controller track the line between current waypoint and previous waypoint. If it overshoots it will just follow the line which is the  best it can do. 

### Changelog Entry
For release notes:
```
Feature/Bugfix XYZ
Documentation: Fixed banking maneuver during a backtransition when the vehicle overshoots the landing point.
```

### Alternatives
We could also instead of checking for an active backtransition, populate the land setpoint type in the triplet and have the controller figure out in that way that it's doing a backtransition.

### Test coverage
SITL so far
### Context
Related links, screenshot before/after, video

### CODE REVIEW

File: src/modules/navigator/mission_base.cpp [hunk 1]
```diff
@@ -564,8 +564,6 @@ void MissionBase::handleLanding(WorkItemType &new_work_item_type, mission_item_s
 			_mission_item.autocontinue = true;
 			_mission_item.time_inside = 0.0f;
 			_mission_item.vtol_back_transition = true;
-
-			_navigator->reset_position_setpoint(pos_sp_triplet->previous);
```
Comment (tstastny) [inline src/modules/navigator/mission_base.cpp:568]: is it possible that a vtol land is triggered without a valid previous waypoint? what will happen then in the FW pos ctrl?

**Review by tstastny — APPROVED** [review]
one minor concern on if VTOL land is triggered without valid previous waypoint -- pls take a look at that comment

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 2]
```diff
@@ -1079,7 +1079,13 @@ FixedwingPositionControl::control_auto_position(const float control_interval, co
...
 	if (_position_setpoint_previous_valid && pos_sp_prev.type != position_setpoint_s::SETPOINT_TYPE_TAKEOFF) {
 		Vector2f prev_wp_local = _global_local_proj_ref.project(pos_sp_prev.lat, pos_sp_prev.lon);
-		navigateWaypoints(prev_wp_local, curr_wp_local, curr_pos_local, ground_speed, _wind_vel);
+
+		if (isDoingBackTransition()) {
+			navigateLine(prev_wp_local, curr_wp_local, curr_pos_local, ground_speed, _wind_vel);
+
+		} else {
+			navigateWaypoints(prev_wp_local, curr_wp_local, curr_pos_local, ground_speed, _wind_vel);
+		}
```
[the remark this answers is not in the document]
Comment (RomanBapst) [reply src/modules/fw_pos_control/FixedwingPositionControl.cpp:1088]: @tstastny What we can do shorterm is maybe to move this code into a `control_backtransition` method? In a follow up PR to this I am actually trying to make navigator set the setpoint type to "LAND" which should give the position controller and idea that we are currently doing a vtol landing.
Then we don't need to mix the logic with the standard waypoint following logic. What do you think?

[the remark this answers is not in the document]
Comment (tstastny) [reply src/modules/fw_pos_control/FixedwingPositionControl.cpp:1088]: > control_backtransition

Could be one way - at least to start splitting out the functionalities.. and then once we can do a larger module split.. the methods are already delineated and ready to migrate

> In a follow up PR to this I am actually trying to make navigator set the setpoint type to "LAND" which should give the position controller and idea that we are currently doing a vtol landing.

I don't fully get what you mean -- but maybe this we could chat offline

> Then we don't need to mix the logic with the standard waypoint following logic. What do you think?

Would be the idea

[the remark this answers is not in the document]
Comment (RomanBapst) [reply src/modules/fw_pos_control/FixedwingPositionControl.cpp:1088]: Currently, the position controller has no clue that he's steering towards a transition waypoint. This is because navigator just publishes a standard position waypoint. If navigator does not do that but just keeps the original intent (which is to land), then the position controller can call the appropriate control functions.

And yes, I agree that the logic should then at some point be pulled into a mode. But as you said, it's clearly easier to do that refactor if the code is already split up nicely in the controller now.

[the remark this answers is not in the document]
Comment (tstastny) [reply src/modules/fw_pos_control/FixedwingPositionControl.cpp:1088]: > Currently, the position controller has no clue that he's steering towards a transition waypoint. This is because navigator just publishes a standard position waypoint. If navigator does not do that but just keeps the original intent (which is to land), then the position controller can call the appropriate control functions.

Ook i understand you now. Yes that would be a nice start!

File: src/modules/navigator/mission_base.cpp (hunk 1 above)
Comment (tstastny) [reply src/modules/navigator/mission_base.cpp:568]: @RomanBapst what about this question?

Comment (RomanBapst) [reply src/modules/navigator/mission_base.cpp:568]: @tstastny Please have a look again. There is now a dedicated method for controlling backtransitions. If the previous waypoint is not valid it will default to waypoint following.

File: src/modules/fw_pos_control/FixedwingPositionControl.hpp [hunk 3]
```diff
@@ -995,7 +1008,8 @@ class FixedwingPositionControl final : public ModuleBase<FixedwingPositionContro
 		(ParamFloat<px4::params::FW_TKO_AIRSPD>) _param_fw_tko_airspd,
 
 		(ParamFloat<px4::params::RWTO_PSP>) _param_rwto_psp,
-		(ParamBool<px4::params::FW_LAUN_DETCN_ON>) _param_fw_laun_detcn_on
+		(ParamBool<px4::params::FW_LAUN_DETCN_ON>) _param_fw_laun_detcn_on,
+		(ParamInt<px4::params::NAV_FORCE_VT>) _param_nav_force_vtol
```
Comment (sfuhrer) [inline src/modules/fw_pos_control/FixedwingPositionControl.hpp:1012]: This could be a bool no?

File: src/modules/navigator/mission_block.cpp [hunk 4]
```diff
@@ -109,8 +115,8 @@ MissionBlock::is_mission_item_reached_or_completed()
 	case NAV_CMD_DO_VTOL_TRANSITION:
 
 		if (int(_mission_item.params[0]) == 3) {
-			// transition to RW requested, only accept waypoint if vehicle state has changed accordingly
-			return _navigator->get_vstatus()->vehicle_type == vehicle_status_s::VEHICLE_TYPE_ROTARY_WING;
+			return _navigator->get_vstatus()->vehicle_type == vehicle_status_s::VEHICLE_TYPE_ROTARY_WING
+			       && !_navigator->get_vstatus()->in_transition_mode;
```
Comment (sfuhrer) [inline src/modules/navigator/mission_block.cpp:119]: Why was this change required? Didn't we on purpose have it accept the transition WP already when the transition was started before?
And I find the comment "transition to RW requested etc" useful, would you mind putting it back? Otherwise one has to look up what "3" stands for.

File: src/modules/navigator/rtl_direct.cpp [hunk 5]
```diff
@@ -279,7 +279,7 @@ void RtlDirect::set_rtl_item()
 			DestinationPosition dest{_destination};
 			dest.alt = loiter_altitude;
 
-			setMoveToPositionMissionItem(_mission_item, dest, rtl_heading_mode);
+			setLandMissionItem(_mission_item, dest, rtl_heading_mode);
```
Comment (sfuhrer) [inline src/modules/navigator/rtl_direct.cpp:282]: This is required such that the type goes to CMD_LAND? Do you see any other side effects from the change? E.g. the heading setpoint could now be different to what it was set previously in setMoveToPositionMissionItem() no?

File: src/modules/navigator/rtl_direct.cpp [hunk 6]
```diff
@@ -310,6 +310,7 @@ void RtlDirect::set_rtl_item()
 			dest.alt = loiter_altitude;
 
 			setMoveToPositionMissionItem(_mission_item, dest, rtl_heading_mode);
+			_navigator->reset_position_setpoint(pos_sp_triplet->previous);
```
Comment (sfuhrer) [inline src/modules/navigator/rtl_direct.cpp:313]: Why is this now required?

**Review by sfuhrer — COMMENTED** [review]
I have some questions in the navigator part, as as we know small changes can cause a lot of ripple effects there.

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 7]
```diff
@@ -2178,6 +2176,42 @@ FixedwingPositionControl::control_manual_position(const float control_interval,
...
+
+	_npfg.setAirspeedNom(target_airspeed * _eas2tas);
+	_npfg.setAirspeedMax(_performance_model.getMaximumCalibratedAirspeed() * _eas2tas);
+
+	if (_position_setpoint_previous_valid && pos_sp_prev.type != position_setpoint_s::SETPOINT_TYPE_TAKEOFF) {
+		Vector2f prev_wp_local = _global_local_proj_ref.project(pos_sp_prev.lat, pos_sp_prev.lon);
+		navigateLine(prev_wp_local, curr_wp_local, curr_pos_local, ground_speed, _wind_vel);
+
+	} else {
+		navigateWaypoint(curr_wp_local, curr_pos_local, ground_speed, _wind_vel);
```
Comment (KonradRudin) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:2197]: With this, it can happen again, that we make a hard turn after we past the waypoit, no? Shouldn't we make it like that: if the current setpoint is a loiter, then navigateLoiter, if it is velocity, then navigatePathTangent, if it is position and has previous navigate line previous to current, and else navigateLine vehicle position to current setpoint position?
I am not sure if this is complete, but i think you want to avoid that mif transitioning it willreverse the direction by passing by.

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 8]
```diff
@@ -708,7 +708,7 @@ FixedwingPositionControl::set_control_mode_current(const hrt_abstime &now)
 
 		} else if (_pos_sp_triplet.current.type == position_setpoint_s::SETPOINT_TYPE_LAND) {
 
-			if (!_vehicle_status.in_transition_mode) {
+			if (!(_vehicle_status.is_vtol && _param_nav_force_vtol.get() > 0)) {
```
Comment (KonradRudin) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:711]: With this, you can enter your mode only if you explicitly want to land and make the back transition. That is the most use case i think, but why not try to make it also when back transitioning mid air?

**Review by KonradRudin — COMMENTED** [review]
What is the benefit of enter the mode already before the transition is happening? Can't we just let it run on normal auto mode until the transition is really triggered?

**Comment by RomanBapst** [comment]
@KonradRudin 

> What is the benefit of enter the mode already before the transition is happening? Can't we just let it run on normal auto mode until the transition is really triggered?

The benefit is that you can have specific logic apply when it's clear that we are heading towards a transition waypoint. Once example would be that you don't want to start loitering even if there is a larger altitude error. This has been an issue in the past.

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 9]
```diff
@@ -708,7 +708,7 @@ FixedwingPositionControl::set_control_mode_current(const hrt_abstime &now)
...
```
[the remark this answers is not in the document]
Comment (RomanBapst) [reply src/modules/fw_pos_control/FixedwingPositionControl.cpp:712]: @KonradRudin I don't understand your point. 

> Now you are dependent, that somewhere exxternally the transition command will be executed

We have always been dependent on someone externally triggering the transition. It's navigator who publishes the transition command the the vtol controller executes. Not that I am a fan of that, but that's how it currently is.

> Why did you change that? I think it is because you want to enter the FW_POSCTRL_MODE_APPROACH_TRANSITON already before the transition happens, but is this really necessary? 

Exactly, I want the controller to be aware of that, because then it can specifically execute. E.g. it will not try to loiter even if there is a large altitude error. Only handling the transition is not enough, because you might have already started loitering. And honestly, I don't want to extend the logic in navigator.

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp (hunk 7 above)
Comment (RomanBapst) [reply src/modules/fw_pos_control/FixedwingPositionControl.cpp:2197]: @KonradRudin Yes, in theory it can. I have verified that it should under normal circumstances not happen when you plan a mission landing or when you use the home land approaches. I am currently actually testing what happens when you use a plan transition command.
I am not opposed to add logic (similar to what we have done on the older releases) which handles that case.

File: src/modules/navigator/mission_block.cpp (hunk 4 above)
Comment (RomanBapst) [reply src/modules/navigator/mission_block.cpp:119]: @sfuhrer Yes, that was the logic before, but I don't it makes sense. Why do you accept a transition when the transition is not done. That's exactly what causes the position controllers and the navigator to be out of sync. Navigator will happily move to the next waypoint while the controller is still dealing with a vehicle in full fw condition.
I am not saying you can't make it work, but I'd rather have everything in sync as good as possible.

Commands are a bit weird in that case.

**Comment by RomanBapst** [comment]
I synced with @KonradRudin on a call and we agreed to the following:
1) The position controller will only switch to the transition mode (APPROACH_TRANSITION currently) when we are actually in transition mode. That way we are sure to execute the correct logic and we are not depending on an external module to issue the transition command.
2) When the vehicle is heading towards the transition point, it needs to be aware that there's potentially a transition coming and thus e.g. should not loiter as a result of an altitude error. We do this by setting the next position setpoint in the triplet to type LAND.
That way the position controller is made aware of the intention to land and can act accordingly.
3) Having a VTOL transition item as part of a normal waypoint mission (not a VTOL LAND waypoint) will not be handled in any special way, e..g the position controller will continue doing normal waypoint following.

File: src/modules/navigator/rtl_direct.cpp (hunk 6 above)
Comment (RomanBapst) [reply src/modules/navigator/rtl_direct.cpp:313]: @sfuhrer Because we don't want line following when moving to the land waypoint during hover.

File: src/modules/navigator/mission_base.cpp (hunk 1 above)
Comment (RomanBapst) [reply src/modules/navigator/mission_base.cpp:568]: @tstastny Should not be an issue anymore, if now the previous waypoint  is invalid, the controller will use the position at the start of the transition and the land point for line following.

File: src/modules/navigator/mission_block.cpp [hunk 10]
```diff
@@ -100,7 +100,13 @@ MissionBlock::is_mission_item_reached_or_completed()
 	// Indefinite Waypoints
 	case NAV_CMD_LAND: /* fall through */
 	case NAV_CMD_VTOL_LAND:
-		return _navigator->get_land_detected()->landed;
+		if (_navigator->get_vstatus()->vehicle_type == vehicle_status_s::VEHICLE_TYPE_FIXED_WING) {
+			// handled below
+			break;
```
Comment (KonradRudin) [inline src/modules/navigator/mission_block.cpp:105]: Are you sure this works? Now in FW the waypoint gets accepted when only horizontally at the landing point. This might be good for VTOL but for FW? It might accept it now before it actually is on the ground, no? Maybe only do this for VTOL land and not for land?

Comment (RomanBapst) [reply src/modules/navigator/mission_block.cpp:105]: @KonradRudin I removed it while doing my investigation for the bug. Did not fix the bug but still will be gone.

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 11]
```diff
@@ -986,8 +983,13 @@ FixedwingPositionControl::handle_setpoint_type(const position_setpoint_s &pos_sp
...
 	const float acc_rad = _npfg.switchDistance(500.0f);
 
-	if (pos_sp_curr.type == position_setpoint_s::SETPOINT_TYPE_POSITION
-	    || pos_sp_curr.type == position_setpoint_s::SETPOINT_TYPE_LOITER) {
+	const bool approaching_vtol_backtransition = _vehicle_status.is_vtol
+			&& pos_sp_curr.type == position_setpoint_s::SETPOINT_TYPE_POSITION && _position_setpoint_current_valid
+			&& pos_sp_next.type == position_setpoint_s::SETPOINT_TYPE_LAND && _position_setpoint_next_valid;
+
+
+	// check if we should switch to loiter but only if we are not expecting a backtransition to happen
```
Comment (RomanBapst) [inline src/modules/fw_pos_control/FixedwingPositionControl.cpp:991]: @KonradRudin 
> Did i miss something, I thought we need now set the next position setpoint triplet to land, in order to detect, that w can ignore the altitude. IS that in and i missed it?

This is where the check is done

**Comment by RomanBapst** [comment]
@KonradRudin @tstastny @sfuhrer Could you guys please update your reviews?

File: src/modules/fw_pos_control/FixedwingPositionControl.hpp [hunk 12]
```diff
@@ -243,6 +246,7 @@ class FixedwingPositionControl final : public ModuleBase<FixedwingPositionContro
 		FW_POSCTRL_MODE_AUTO_PATH,
 		FW_POSCTRL_MODE_MANUAL_POSITION,
 		FW_POSCTRL_MODE_MANUAL_ALTITUDE,
+		FW_POSCTRL_MODE_APPROACH_TRANSITON,
```
Comment (KonradRudin) [inline src/modules/fw_pos_control/FixedwingPositionControl.hpp:249]: Cosmetic: Could rename it to MODE_BACK_TRANSITION

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp (hunk 11 above)
Comment (KonradRudin) [reply src/modules/fw_pos_control/FixedwingPositionControl.cpp:991]: That is not what i meant, sorry, i saw that we correctly set the next position setpoint to land. but within the https://github.com/PX4/PX4-Autopilot/blob/5aa2518f8cb0277c6036efbf19f7719d529bc143/src/modules/navigator/mission_block.cpp#L65 function, it still only accept it when both within the horizontal and altitude acceptance. You might need to tweak this function here https://github.com/PX4/PX4-Autopilot/blob/5aa2518f8cb0277c6036efbf19f7719d529bc143/src/modules/navigator/navigator_main.cpp#L1108-L1109
To set altitude acceptance to infinite when  we know that the backtransition is happening.
Thinking more about it and seeing, that in fixed wing case, we actually want to make the opposite, maybe we really need to add vtol_back_transition flag from the mission_item also in the setpoint triplet, so we know, that we can ignore altitude.

Comment (RomanBapst) [reply src/modules/fw_pos_control/FixedwingPositionControl.cpp:991]: @KonradRudin Maybe I'm missing something but we do set the altitude acceptance radius to infinity if the mission item has the flag `vtol_back_transition` set to true.
https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/navigator/mission_block.cpp#L357C5-L357C18

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp [hunk 13]
```diff
@@ -986,8 +983,13 @@ FixedwingPositionControl::handle_setpoint_type(const position_setpoint_s &pos_sp
...
+	if (pos_sp_curr.type == position_setpoint_s::SETPOINT_TYPE_POSITION && !approaching_vtol_backtransition) {
```
[the remark this answers is not in the document]
Comment (RomanBapst) [reply src/modules/fw_pos_control/FixedwingPositionControl.cpp:992]: @KonradRudin The reason I added that is because we have two separate modules (navigator & position controller) which could use different horizontal acceptance radius and therefore not be "in sync" anymore. I agree that in theory now this should not be needed as I assume navigator and the position controller use the same value.
It's a protection in the sense that if we know we are expecting a backtransition, I don't ever want it to loiter.
The only bad case I see is if for some reason the logic in navigator for switching to the next waypoint would not trigger, but that would mean that the vehicle is blown away so much that the distance to the land waypoint does not reach the minimum required.
That could in theory happen when you have extremely strong wind and thus a very low groundspeed. That would make the transition distance very small and in theory it would be possible to "miss" the transition point.
We could add a minimum transition distance to make sure that risk is lower.
What do you think?

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp (hunk 11 above)
Comment (RomanBapst) [reply src/modules/fw_pos_control/FixedwingPositionControl.cpp:991]: @KonradRudin I made a quick test and increased the altitude of the mission item when the vehicle is approaching the transition by a value of 1000m. To me everything looks nominal, navigator accepts the mission item despite the huge altitude error. Is this what you meant?

image

Comment (KonradRudin) [reply src/modules/fw_pos_control/FixedwingPositionControl.cpp:991]: ITs fine, i missed the line you pointed out to me, thanks, and sorry!

File: src/modules/fw_pos_control/FixedwingPositionControl.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (KonradRudin) [reply src/modules/fw_pos_control/FixedwingPositionControl.cpp:992]: Well, we need to think of what is the best thing to do when we reached the end waypoint and the back transition has not triggered yet. Before it was loiter, now it the "flower" pattern, both are not good. But i get it, that in theory it should only happen when we are blown away so flying to the pattern might be better. Still hate this obfuscating navigator <-> controller interaction and should really kill current navigator once to make this better. But let's get it in this way.

**Review by KonradRudin — APPROVED** [review]
